### ISSUE TITLE
This document is published but its url cannot be routed 

### ISSUE THREAD

**Issue by sbosell** [umbraco/Umbraco-CMS#9514]
Sometimes a node in umbraco will be in a state of published but not routable when this should not be allowed to occur.  Umbraco in this case throws a 404 for that route.

There are several cases of this being [reported](https://our.umbraco.com/forum/using-umbraco-and-getting-started/104334-this-document-is-published-but-its-url-cannot-be-routed) as well as other issues reported here ([duplicate](https://github.com/umbraco/Umbraco-CMS/issues/9523))

Related to issue #7575 

## Umbraco version

I am seeing this issue on Umbraco version: 8.6.2, 8.7.x, and 8.9.1

Reproduction
------------

I have a process that syncs data in the backend every 30 minutes to a node's descendants and then the content api calls SaveAndPublish and/or SaveAndPublishBranch (tested with both in the sync).  The sync uses the content api via a recurring task, all standard features of Umbraco.  In my case we have no cultures defined and only one root site.   The most basic Umbraco setup.  Others are reporting the same issue so it may have nothing to do with the Content Api.

About 1 time per day the root node of the sync will be in this state of the url can't be routed and it is random in that it only occurs every once in a while which leads me to believe there is a locking mechanism with how the cache works.  This should never happen, there should be no code path that unpublished or deroutes a node like this in the save and publish.

Hosted on Azure in a single site (not load balanced) with all the appropriate settings set via the Umbraco [Documentation](https://our.umbraco.com/documentation/getting-started/setup/server-setup/azure-web-apps).

### Bug summary

A node becomes not routable when this should never happen.

### Specifics

I can privately provide a URL 

### Steps to reproduce

It happens randomly

### Expected result

A valid published node in Umbraco should NEVER be in a state of the URL not being routable. 

### Actual result

Umbraco backend reports the URL not being routable and the node is effectively unpublished.

---
_This item has been added to our backlog [AB#9840](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/9840)_

**Comment by sbosell** [comment]
This error just occurred right now.  

![image](https://user-images.githubusercontent.com/1896962/101559516-483d1000-398f-11eb-8cad-20492b2abe89.png)

**Comment by sbosell** [comment]
This happened almost daily this week and is being reported by others.  Duplicate ticket: https://github.com/umbraco/Umbraco-CMS/issues/9523

![image](https://user-images.githubusercontent.com/1896962/102013236-d6f9b600-3d1c-11eb-818a-fcf1c739f795.png)

**Comment by Jay-umbr** [comment]
Seen several cases of this, too. Could not nail down the cause - seems to happen totally randomly, occurred sometimes after a deployment, sometimes after a content transfer on Cloud, sometimes just after leaving the site be and coming back to it after a few days.
This happens regardless of models builder mode (had it happen on sites using PureLive and AppData), and on plain/package-less Umbraco installations.

**Comment by sbosell** [comment]
This continues to happen almost daily.   Umbraco randomly unpublishes nodes and the content is 404 not found.

**Comment by Nicholas-Westby** [comment]
FYI, I've seen this with an Umbraco 8.6.2 install, so it has been happening at least since then.

**Comment by sbosell** [comment]
@Nicholas-Westby Thanks. I updated the ticket description to include 8.6.2.

**Comment by kjac** [comment]
In the related issues someone noted that they could fix the broken nodes by rebuilding the cache. I wonder if disabling the cache DB (thus forcing it to rebuild upon startup) is a workaround for this?

If anyone should be inclined to test this, try adding the following to an `IUserComposer`:

```cs
public void Compose(Composition composition)
{
  composition.Register(factory => new PublishedSnapshotServiceOptions
  {
    IgnoreLocalDb = true
  });
  // ...
}
```

**Comment by sbosell** [comment]
@kjac what related issue are you referring to?

I can test the change this week.

**Comment by sbosell** [comment]
I deployed the change this morning and the documentation for it is [here](https://our.umbraco.com/documentation/getting-started/setup/server-setup/load-balancing/azure-web-apps).  If the only consequence is a slower startup time on a new server I can live with that if it addresses this problem.  I should point out we do not have replica servers.

**Comment by sbosell** [comment]
The change made no difference as we experienced the same issue this afternoon.

**Comment by Shazwazza** [comment]
Do you have multiple root nodes? And what domains do you have assigned to them? If you have multiple root nodes, only one of them can have no domains, all other ones must have domains.

**Comment by sbosell** [comment]
I have the simplest setup of one root and one domain.

**Comment by vaags** [comment]
I'm seeing this problem on Umbraco 8.6.1. There's only one root node, but there are multiple language variants. On several pages, only one of the langauges is routable, despite all of them being published.

**Comment by nikolajlauridsen** [comment]
It would be helpful if we can determine if it's caused indirectly by some action in the backoffice, or by some background task or something similar.

Do you know if the issue occurs at any time of the day, does it happen in the middle of the night when no one is using the back office as well?

Would it be possible for you to check the umbracoAudit table in the database as well? Maybe something happens leading up to the bug that might be able to help us.

**Comment by sbosell** [comment]
@nikolajlauridsen

The issue for us happens randomly and there are very few updates in the backoffice (maybe one time per week at the most).  We do have a process that runs every 30 minutes that's outlined in the ticket.  Here are some of the times of day it has occurred:

* 2021-01-07 14:30:25
* 2021-01-05 17:48:25
* 2020-12-28 18:23:16
* 2020-12-28 17:50:30
* 2020-12-25 13:57:00
* 2020-12-25 01:29:39
* 2020-12-24 20:54:19
* 2020-12-24 06:42:08
* 2020-12-22 21:36:54

[Here ](https://docs.google.com/spreadsheets/d/1mf8P3MfRzr04H1Cec-JkfYfIFQMpLC_GcbbG31Hueo8/edit?usp=sharing)is the full audit log.

**Comment by kimschurmann** [comment]
Could this issue please get some attention - we do not dare to upgrade to the next version from 8.5.5. We are kinda stuck - there are issues in 8.5.5 that are fixed in the next versions, but if we do - we risk new and more serious issues :sigh:

**Comment by bergmania** [comment]
We are continuously looking at this issue. Sadly we can't reproduce and we are therefore looking for a needle in a haystack.

@vaags and @sbosell, when you say "one domain" are you talking about domains/hostnames in Umbraco or did you not add any domains in "Culture and Hostnames"

![image](https://user-images.githubusercontent.com/1561480/104564316-0f0d5500-564b-11eb-96e2-1e1bdf0cb44f.png)

**Comment by sbosell** [comment]
@bergmania in my case we have english as the only language and we actually have domain.com And www.domain.com setup on that screen. I'm happy to share the backend info privately with someone from the Umbraco team.

**Comment by orsinic** [comment]
We also see nodes randomly losing the url ("This document is published but its url cannot be routed").

In our case we see it already for several months on two separate Umbraco Cloud projects with always the latest version (so the current latest version has the issue, but also several earlier versions). Unfortunately I cannot tell since which version exactly this started to occur.

It only seems to happen after a (code) deployment from a lower environment. I haven't seen this occur randomly without a deployment.

Both projects are setup with multiple subsites at the top level, each with their own domain configured in "Culture and Hostnames". For both projects multiple languages are used but in a v7 way (eg not making use of v8 variants).
One of the projects is a baseline project (where we see this behavior on all child projects) and the other is a plain/single project with two environments.
In one project we have jobs running that create content using the content service (similar to some other cases mentioned here). But in the other project we don't create content programmatically.

Rebuilding the cache fixes the issue for us. So unfortunately this is now part of our workflow after each deployment.

Shared info about this with @Jay-umbr via Umbraco Cloud support previously. If needed I can share extra information to find the root cause of this.

**Comment by sbosell** [comment]
@orsinic I can fix the issue by rebuilding the cache or by publishing the node that was derouted. It happens randomly for us (self hosted azure). Derouting shouldn't even be possible.

**Comment by kimschurmann** [comment]
Good to know there is a work around if the issue occurs - but I don't think I will upgrade just yet if a part of the deployment rutine is to rebuild cache of republish nodes after each deployment.

**Comment by bergmania** [comment]
For those of you, with only a single hostname assigned, could you try to remove the hostname and see if the error goes away. There's not really a reason to have just one hostname with one root node. If that 'fixes' the issue, then we know its a domain issue. That would help a lot.

**Comment by Zweben** [comment]
I was directed here by Umbraco support after encountering an issue with content suddenly disappearing from our production site the second time this month. The two times we had issues had similar circumstances and resolutions, but presented as different front-end problems, so I'll list them separately:

Earlier instance:
* Happened right after a deployment from Dev to Live in Umbraco.io
* Happened very soon after our upgrade to 8.9.1
* All text stored in Rich Text Editors disappeared from the front end
* I don't recall if pages went unpublished as well, but I don't think so
* Was resolved after rebuilding the DB cache and re-publishing all pages

Latest instance:
* Happened right after a deployment from Dev to Live in Umbraco.io
* No upgrades recently
* No text property content disappeared from pages, but many pages disappeared from C# query results in the front-end
* Was resolved after rebuilding the DB cache and re-publishing all pages

Info about the site:
* Uses a custom Content Finder
* Uses multiple root nodes
* Uses multiple languages, with multi-lingual properties on most doctypes

Errors I am seeing in the logs:
* From before the latest issue (seems unrelated): 
`System.IO.FileNotFoundException: Could not load file or assembly 'Umbraco.Cloud.StorageProviders.AzureBlob' or one of its dependencies. The system cannot find the file specified.`
* From right before the issue started:
`System.Data.SqlClient.SqlException (0x80131904): Lock request time out period exceeded.`
* While the issue was affecting the site, as a result of content queries:
`System.InvalidOperationException: Sequence contains no elements`
and `System.NullReferenceException: Object reference not set to an instance of an object.`
* Few errors in the log since the nodes were restored to their published state.

**Comment by bergmania** [comment]
Hi @Zweben.. Thansk for the detailed description. Do you have the stacktrace from the following exception `System.Data.SqlClient.SqlException (0x80131904): Lock request time out period exceeded.`

**Comment by bergmania** [comment]
When anyone is seeing this issue next time. Please go the content in backoffice and post the response of this call here: `{domain}/umbraco/backoffice/UmbracoApi/Content/GetById?id={contentId}`. I don't understand how the link from @sbosell can be shown at the same time as the message..

**Comment by sbosell** [comment]
@bergmania 

The api call you asked for is attached to the ticket.  I anonymized three entries in the file. Owner names and the domain was changed to mydomain.com instead of the actual domain.    Also attached is a screenshot of the backend.

![image](https://user-images.githubusercontent.com/1896962/104827644-f7da8d00-582d-11eb-8cbf-968ea368cecb.png)
[GetById.txt](https://github.com/umbraco/Umbraco-CMS/files/5825324/GetById.txt)

**Comment by Zweben** [comment]
> 
> 
> Hi @Zweben.. Thansk for the detailed description. Do you have the stacktrace from the following exception `System.Data.SqlClient.SqlException (0x80131904): Lock request time out period exceeded.`

I do not. This has only happened on our Live site recently, where debug mode is off. Is there a way I can get a stack trace without debug mode / detailed error pages enabled?

**Comment by bergmania** [comment]
> I do not. This has only happened on our Live site recently, where debug mode is off. Is there a way I can get a stack trace without debug mode / detailed error pages enabled?

The stack trace should still be part of the log entry in the log file..

**Comment by bergmania** [comment]
Thanks @Zweben.. Both of these seems to happen from an user interaction in backoffice, where a content is published. Sadly we don't have info in the stacktrace, that could tell if it was the same page that later becomes unavailable.

**Comment by kimschurmann** [comment]
I think we just experienced this problem i 8.5.5 - we busted the caches and it worked again. We have two root nodes.

**Comment by bergmania** [comment]
> I think we just experienced this problem i 8.5.5 - we busted the caches and it worked again. We have two root nodes.

Interesting, it looks like it's a problem that has been here super long. Unfortunately, it just does not make it easier to find the root cause.. We are still searching for the issue, without being able to reproduce.

**Comment by kimschurmann** [comment]
Honestly I dont think its a problem related to multiple root nodes - but not sure :)

**Comment by bergmania** [comment]
> Honestly I dont think its a problem related to multiple root nodes - but not sure :)

That was just one question to narrow down the issue

**Comment by Shazwazza** [comment]
@sbosell re: the above question from @bergmania 

> For those of you, with only a single hostname assigned, could you try to remove the hostname and see if the error goes away. There's not really a reason to have just one hostname with one root node. If that 'fixes' the issue, then we know its a domain issue. That would help a lot.

Are you able to try without having a hostname assigned to your single root node? We want to see if this has anything to do with domain caches.

**Comment by Shazwazza** [comment]
@sbosell you mentioned in #9523 that this happens every day for you which is interesting since most others are very sporadic/random. As we need to try to reproduce but we can't, seems like you have are able to best assist. For more logging output that may help, you can add this line to your serilog.config file:

```xml
<add key="serilog:minimum-level:override:Umbraco.Web.Routing.PublishedRouter" value="Debug" />
```

This will log A LOT of info but if this happens everyday for you, perhaps you can enable this for a day until it happens and share the log outputs?

**Comment by sbosell** [comment]
@Shazwazza  I enabled the serilog setting.  As long as it doesn't negatively impact the site it isn't a problem.  Here is a list of times it has happened over the last week or so:

![image](https://user-images.githubusercontent.com/1896962/105704674-27982d80-5edd-11eb-8877-96d272d2a6cd.png)

**Comment by sbosell** [comment]
@Shazwazza Is there a private way I can share the file with the team?

**Comment by bergmania** [comment]
> @Shazwazza Is there a private way I can share the file with the team?

@sbosell, feel free to send a mail to me, then I will distribute it internally to those to help on this task.
bmb@umbraco.dk

**Comment by orsinic** [comment]
@sbosell quick question, how do you get a list of all nodes having the "This document is published but its url cannot be routed" issue?

As noted before (and as far as I know) we do not randomly get this, but only after code deployments. Being able to have a report with the impacted nodes might help digging a bit deeper or trying to get this to be reproduced.

Thanks in advance!

**Comment by sbosell** [comment]
@orsinic I have an uptime monitor on every page on the site but we only see the issue on one page so I immediately know when it is down. Youc an see that in the graphic posted above in my last message.

@bergmania Sent you the file.

**Comment by Shazwazza** [comment]
@sbosell thanks for enabling and sending those logs. [here](https://github.com/umbraco/Umbraco-CMS/issues/9523#issuecomment-744704665) you've mentioned you use a custom UmbracoVirtualNodeRouteHandler. Based on your logs I'm guessing that the URL that stops working is based on this custom UmbracoVirtualNodeRouteHandler? If you can confirm - then this is the problem area for your specific issue and will help with investigation.

@Zweben [here](https://github.com/umbraco/Umbraco-CMS/issues/9514#issuecomment-760387953) you mention you use a custom IContentFinder. Do you also have a custom url provider? Is it possible to disable these to see if the error goes away? You also mention you use HangFire  - what and when exactly is this doing? You've also mentioned `Uses multiple root nodes, one with hostnames assigned, three without` ... in theory there should only ever be one root node without a domain else there can be problems with inbound routing a naming conflicts which can end up resulting in routing errors such as these. Can you please ensure there is only one root node without a domain (else you can have all root nodes with domains too) and see if that resolves the issue?

There are a lot of varying setups, questions and issues on this one thread and I don't really think they are all caused by the exact same thing. There's also a few suggestions above for things to try that we haven't heard back about yet. To recap:

* If you only have a single root node with a domain assigned - try un-assigning the domain since that isn't necessary with this setup and see if the issue resolves itself... this would mean that it's an issue with domains and will give as a place to investigate.
* If you have multiple root nodes - only one of those root nodes can have no domains assigned. Ensure this is the case and please report back if that solves your issue. 
* If you have custom extensions in place like IContentFinder, UrlProviders and/or custom routing with things like UmbracoVirtualNodeRouteHandler ... if its possible to disable these and see if the problem goes away then it will give us a place to start looking to replicate (which may require the code from these extensions to try to replicate)

**Comment by sbosell** [comment]
@Shazwazza - The node that is being derouted does not have a custom UmbracoVirtualNodeRouteHandler, but some of its children do.  For instance the route in my case that is derouted is for the /locations node and all the custom route handlers are for /locations/{id}, etc but not ever for /locations.

**Comment by benbracedigital** [comment]
We're getting the same issue as this with many pages on a site we look after. The site is running 8.9.1. We noticed this a couple of months ago and put it down to the cache need rebuilt and/or page published.

We also thought it was a conflict between controller names with doc types (even though they were surface controllers) so we changed the name of those classes but it didn't help.

This is now happening several times a day and causing big issues for our client. The site is hosted on a virtual server (not azure or cloud based) and is not load balanced.

Any update would be much appreciated.

**Comment by MartinThomasCW** [comment]
Same here - we're on 8.6.6.   Originally on 8.6.1 we assumed we just had an issue with the indexes (fixed in either 8.6.2 or 8.6.3), but that was just masking this problem.   As per benbrace, this is a real issue for our client as the area of the site that's disappearing is the section where people have paid good money to access the content as members.

It happens sporadically - sometimes multiple times daily, then nothing for a week etc.   There's no real discernible pattern - it happens when the site has users and when there's no traffic.   With and without content editors changing / uploading content etc.

It's a single root node site, single language with no exotic elements at all.  We did try and think laterally to temporarily work out the issue, by installing hangfire, detecting the page going down and automating an http post to the /BackOffice/Api/NuCacheStatus/ReloadCache endpoint, but to no avail.  The call returns http 200, but i suspect the security context is missing to authenticate the hangfire user as admin etc.   (This issue was happening well before hangfire was installed btw - we only installed it as a last gasp attempt)

If you need any information (nucache dumps, access to the site etc) please don't hesitate to reach out - the sooner we get this sorted the better for the whole community!

Many thanks

**Comment by Zweben** [comment]
> @Zweben [here](https://github.com/umbraco/Umbraco-CMS/issues/9514#issuecomment-760387953) you mention you use a custom IContentFinder. Do you also have a custom url provider? Is it possible to disable these to see if the error goes away? You also mention you use HangFire - what and when exactly is this doing? You've also mentioned `Uses multiple root nodes, one with hostnames assigned, three without` ... in theory there should only ever be one root node without a domain else there can be problems with inbound routing a naming conflicts which can end up resulting in routing errors such as these. Can you please ensure there is only one root node without a domain (else you can have all root nodes with domains too) and see if that resolves the issue?

@Shazwazza: I don't remember exactly, but I think my IContentFinder doesn't have a custom URL provider that corresponds to it (it's a bit of an unusual case), but I believe I do have a custom URL provider used elsewhere. Unfortunately, I'm not able to disable either as they're both being used in production.

Hangfire is running a few tasks: importing JSON records into nodes, and sending out a few email reports. I've checked them fairly recently and they weren't hitting any errors and didn't appear to be causing any issues.

Regarding the multiple root nodes, I wasn't aware of that... there should probably be some warning in the CMS, as it's quite easy to configure things this way. I should be able to move things around so that there is only one root node without a domain assigned. Is it safe to have a second set of domainless nodes in the root if that second set is purely "settings" nodes that have no template assigned and are therefore not routed to?

**Comment by sbosell** [comment]
Does anyone know if you can inject an IPolicyCache into an UmbracoApiController like the following and will clearing it cause the nucache to rebuild?

```public ResyncController : UmbracoApiController
  private readonly IAppPolicyCache _runtimeCache;
        public ResyncController(IAppPolicyCache cache)
        {
            _runtimeCache = cache;
             // this would be called in anohter method just putting here for this post
            _runtimeCache.Clear();
        }
```

**Comment by Shazwazza** [comment]
> Does anyone know if you can inject an IPolicyCache into an UmbracoApiController like the following and will clearing it cause the nucache to rebuild?

No, that is not how nucache works

**Comment by Shazwazza** [comment]
> Regarding the multiple root nodes, I wasn't aware of that... there should probably be some warning in the CMS, as it's quite easy to configure things this way. I should be able to move things around so that there is only one root node without a domain assigned. Is it safe to have a second set of domainless nodes in the root if that second set is purely "settings" nodes that have no template assigned and are therefore not routed to directly?

The reason why multiple root nodes without domains is not supported is because you can easily end up with ambiguous route URLs based on the names of nodes. This isn't guaranteed but it can happen (have seen it often) and there is no guarantee that it will pick the 'first' ambiguous URL over the 2nd or 3rd. This has previously resulted in nodes going 'missing' until that URL cache was cleared. You can always assign dummy domains to settings root nodes, etc... This also depends on if your settings nodes have templates and/or can be routed to, what their names are , etc... My point is, that it's impossible to understand everyone's particular website so the general rule is - do not have more than one root node without a domain = safe.

**Comment by Shazwazza** [comment]
@MartinThomasCW this point is also interesting:

> As per benbrace, this is a real issue for our client as the area of the site that's disappearing is the section where people have paid good money to access the content as members.

It seems peculiar that only this section does this. I know this thread is super long but as I've mentioned above I suspect that there can be multiple different issues here because each person's site is different. Is there something special about this section? As above, do you have custom IContentFinder, UrlProviders, UmbracoVirtualNodeRouteHandler, etc... ?

**Comment by Shazwazza** [comment]
@sbosell This how the nucache controller works in the back office: https://github.com/umbraco/Umbraco-CMS/blob/v8/contrib/src/Umbraco.Web/Editors/NuCacheStatusController.cs#L60 ... if you need to currently need to resort to brute force techniques

**Comment by MartinThomasCW** [comment]
@Shazwazza - nothing in particular about the pages that clash - they are publicly accessible, but serve as the landing pages to the premium content.   Their children are protected content (secured with standard role based access checks).  Recently there has also been another page in the non "premium" section that got un-routed.  Literally nothing special about that page that i can tell either.

Worth mentioning that I have executed your CheckAndFixPaths sql script, but that returned no results.  Good in one regard, but it's the hope that kills you when you're looking for a fix! :)

**Comment by benbracedigital** [comment]
@Shazwazza We have no IContentFinder, UrlProviders, UmbracoVirtualNodeRouteHandler on our site. The pages which this affects is completely random. Some are one level deep, some are two or three levels deep. From an Umbraco perspective, the affected pages are doing nothing special. They call SurfaceControllers to render forms but don't even have their own RenderControllers.
Today I'm setting up uptime alerts on a lot of the pages to get a good understand of how often they go down and which pages as currently we're relying on our client and users telling us when pages are down which isn't a good look.

**Comment by MartinThomasCW** [comment]
Apologies @Shazwazza , totally forgot to answer your main question!  No, we don't have any custom IContentFinder, UrlProviders or UmbracoVirtualNodeRouteHandlers.

**Comment by ronaldbarendse** [comment]
> In the related issues someone noted that they could fix the broken nodes by rebuilding the cache.

I've seen similar issues after changes were deployed using Umbraco Deploy (on the live environment on Umbraco Cloud): it randomly breaks the content cache and manually rebuilding the cache is the only way to fix this. This also happens on very simple sites (using culture variants with only a path configured on the single root node, e.g. `/en` and `/nl` for the respective languages).

One of the things I noticed in the logs is Umbraco Deploy suspending the document cache (message: `Suspend document cache.`, context: `Umbraco.Web.Suspendable+PageCacheRefresher`) and later resuming it without reloading (message: `Resume document cache (reload:False).`). And whenever it has happened, I also noticed the caches had multiple generations and/or snapshots: rebuilding the cache always resulted in 1 generation and 1 snapshot).

**Comment by ronaldbarendse** [comment]
I've found these steps (while debugging another issue: https://github.com/umbraco/Umbraco-CMS/issues/9751) that are reproducable:
- Remove allowed template(s) from a document type
- Revert the change in document type UDA file and force a deploy (using the marker file `data\deploy`)
- All content using this document type now result in a 404 and show the 'url cannot be routed' error in the back-office
- The published cache status shows: `Database cache is ok. ContentStore contains 23 items and has 2 generations and 1 snapshot. MediaStore contains 20 items and has 1 generation and 1 snapshot.`

There are 2 ways to fix the content cache (and make the content routable again):
1. Publishing the root node with all descendants: this increased the cache generation count
2. Reloading the memory cache: this updated the generation count back to 1 (and if it didn't, a 'collect' would)

Also note: when the cache was in this corrupt state, only publishing the unroutable content increased the generation count without fixing the issue.

**Comment by nikolajlauridsen** [comment]
Thanks a lot for your steps to reproduce @ronaldbarendse. I've tried to follow your instruction, but unfortunately with no luck, I've made a little video of my attempt, if you have a minute to check if I'm misunderstanding your steps that would be awesome.

I'm no video maker expert but hopefully, it's understandable.


https://user-images.githubusercontent.com/16456704/106766856-8f1b4f00-663a-11eb-9a45-43cbaad83532.mp4

**Comment by kimschurmann** [comment]
This is starting to become a serious issue for us - any progress or work being done? Still umbraco 8.5.5.

**Comment by kimschurmann** [comment]
Any hack or work around tip would be appreciated...

**Comment by bergmania** [comment]
@kimschurmann, We are still investigating and work on this task. We still need a solid way to reproduce it, so any help is welcome. 

Especially feedback on @nikolajlauridsen's video would be interesting.

**Comment by Shazwazza** [comment]
Hello everyone, 

Can anyone/everyone tell me if they have this package installed or not? https://our.umbraco.com/packages/website-utilities/virtualnodes/ 

This is one of causes of this error and will occur depending on how your site is setup and how you have configured this.

**Comment by benbracedigital** [comment]
@Shazwazza We do NOT have this installed.

**Comment by MartinThomasCW** [comment]
Nope, nor do we, sorry

**Comment by sbosell** [comment]
I don't have it installed either.

**Comment by Nicholas-Westby** [comment]
Same, don't have VirtualNodes installed.

**Comment by Zweben** [comment]
Still having pretty frequent issues with the homepage effectively going unpublished immediately after a deployment.

No VirtualNodes package installed, or any other Umbraco packages for that matter. I'm also working towards getting our site down to one root node, to see if that improves anything.

**Comment by Shazwazza** [comment]
Thanks all.

@Zweben it is very difficult here to track what everyone is doing and have tried. I mentioned above about how root nodes and domains work. You don't need to have a single root node to test, you just need to have domains properly assigned as to avoid any ambiguous routes, see https://github.com/umbraco/Umbraco-CMS/issues/9514#issuecomment-770515412. But I would be interested to see if the problem still exists with a single root node. Right now any testing, help, feedback, etc... is most welcome to move the investigation along. 

This is how you would setup a tree to cause the ambiguous routes issue - none of these nodes have domains and `HideTopLevelNodeFromPath` = true (the default). If you setup a tree like this then "Home" will be ambiguous because it's trying to make the URL `/home` for both nodes and there's no guarantee which Home will be the matching one. 

* Brand A
  * Home
* Brand B
  * Home

_However_, in this case I don't think that ambiguous routes and missing domains is the issue because testing this, the error message in the UI is this: "This document is published but its URL would collide with content /Brand A/Home (id=123456)"

That said - the way to avoid this and possibly other routing issues is to ensure that domains are assigned (even if they are dummy domains for settings nodes, etc...) at the root level nodes if multiple root nodes exist. It is still worth trying this if you don't have this setup to see if helps.

This particular error message: "This document is published but its url cannot be routed" is when no content can be resolved to the URL that's being tried. Unfortunately there is no logging that occurs when this happens. I've created a PR which will hopefully be in the next patch version which will help shed some light on this hopefully: https://github.com/umbraco/Umbraco-CMS/pull/9830/files#diff-cd06d5e77b193e5227cafc506b674f6620ba39c5c4f36acd55439db48ae573a0R195 

There is certainly some different issues at play here and each persons environment/setup will probably have some varying thing. For example, one of the issues for some was the VirtualNodes package.

**Comment by Shazwazza** [comment]
@MartinThomasCW  is your site an Umbraco Cloud site? If so, would it be possible to invite me as an admin to help investigate? (shannon at umbraco dot com).  And/or if at all possible to enable this verbose logging in the serilog.config file: `<add key="serilog:minimum-level:override:Umbraco.Web.Routing.PublishedRouter" value="Debug" />` just so that it's enabled when this issue happens and then it can be turned off again. It's quite verbose so will log a lot. If that's possible could you email me those logs?

**Comment by nzdev** [comment]
As a note as I have experienced some issues perhaps related, I've seen ObjectDisposedException occur for the snapshot class in nucache when trying to route. Also if the routes collide from two root nodes I've also seen this sort of issue.

**Comment by Shazwazza** [comment]
@nzdev if you have, remember or gather any further bits of information on this please let me know (stack traces, etc...)

**Comment by MartinThomasCW** [comment]
@Shazwazza it's not umbraco cloud, but i'll check with the client for access - not a problem as far as i'm concerned though.   Logging added and going out imminently (they were due a release for other fixes so good timing!)

**Comment by Zweben** [comment]
> @MartinThomasCW is your site an Umbraco Cloud site? If so, would it be possible to invite me as an admin to help investigate? (shannon at umbraco dot com). And/or if at all possible to enable this verbose logging in the serilog.config file: `<add key="serilog:minimum-level:override:Umbraco.Web.Routing.PublishedRouter" value="Debug" />` just so that it's enabled when this issue happens and then it can be turned off again. It's quite verbose so will log a lot. If that's possible could you email me those logs?

Ours is an Umbraco Cloud site; I could probably invite you to the project if you think it would help. I'll also try to remember to turn on verbose logging next time I publish something to Live.

**Comment by sbosell** [comment]
My site is hosted in an Azure Web App in the simplest way without any scale out (not Umbraco Cloud).  Is everyone else hosted on Azure or does anyone here self host?

**Comment by Nicholas-Westby** [comment]
@sbosell We host ourselves (in an AWS VM, which AWS refers to as EC2).

**Comment by benbracedigital** [comment]
We host the site on our own VPS. Not Umbraco Coud or other cloud such as Azure or AWS

**Comment by sbosell** [comment]
@benbracedigital @Nicholas-Westby  I thought this might have been related to how Azure Web App's manage their underlying infrastructure (nodes spin up and down and connect to a virtual file system) but if you are self-hosting on a Virtual Machine that points to something more general in my mind.  I went through the code for when a route is looked up but I don't think that's the issue, something in the cache layer I think is derouting these nodes in some way.

**Comment by Shazwazza** [comment]
@Zweben 

> I could probably invite you to the project if you think it would help. I'll also try to remember to turn on verbose logging next time I publish something to Live.

Yes it would certainly help me if you could invite me that would be great and also let me know if/when you enable that logging. It only needs to be enabled before and up to when you experience this issue and then can be disabled again.

**Comment by Shazwazza** [comment]
Hi all - another question for everyone: For any items that become un-routable, are you using any of the built in 'special' umbraco route property types: `umbracoUrlAlias`, `umbracoUrlName`, `umbracoRedirectId` or `umbracoRedirect`?

**Comment by nzdev** [comment]
Getting the error below @Shazwazza.

Somehow the RouteCollection is getting duplicate and null values.

```
Value cannot be null.
Parameter name: item

   at System.Web.Routing.RouteCollection.InsertItem(Int32 index, RouteBase item)
   at System.Web.Mvc.RouteCollectionExtensions.FilterRouteCollectionByArea(RouteCollection routes, String areaName, Boolean& usingAreas)
   at System.Web.Mvc.RouteCollectionExtensions.GetVirtualPathForArea(RouteCollection routes, RequestContext requestContext, String name, RouteValueDictionary values, Boolean& usingAreas)
   at System.Web.Mvc.UrlHelper.GenerateUrl(String routeName, String actionName, String controllerName, RouteValueDictionary routeValues, RouteCollection routeCollection, RequestContext requestContext, Boolean includeImplicitMvcValues)
   at System.Web.Mvc.UrlHelper.Action(String actionName, String controllerName, RouteValueDictionary routeValues)
   at Umbraco.Web.UrlHelperExtensions.GetUmbracoApiService(UrlHelper url, String actionName, Type apiControllerType, RouteValueDictionary routeVals) in D:\a\1\s\src\Umbraco.Web\UrlHelperExtensions.cs:line 79
   at Umbraco.Web.UrlHelperExtensions.GetUmbracoApiServiceBaseUrl[T](UrlHelper url, Expression`1 methodSelector) in D:\a\1\s\src\Umbraco.Web\UrlHelperExtensions.cs:line 44
   at Umbraco.Web.Editors.BackOfficeServerVariables.GetServerVariables() in D:\a\1\s\src\Umbraco.Web\Editors\BackOfficeServerVariables.cs:line 106
   at Umbraco.Web.Editors.BackOfficeServerVariables.BareMinimumServerVariables() in D:\a\1\s\src\Umbraco.Web\Editors\BackOfficeServerVariables.cs:line 69
   at Umbraco.Web.HtmlHelperBackOfficeExtensions.BareMinimumServerVariablesScript(HtmlHelper html, UrlHelper uri, String externalLoginsUrl, UmbracoFeatures features, IGlobalSettings globalSettings) in D:\a\1\s\src\Umbraco.Web\HtmlHelperBackOfficeExtensions.cs:line 45
   at ASP._Page_umbraco_Views_Default_cshtml.Execute() in C:\Users\chad.currie\source\repos\trn-umbraco\NZME.Radio.Web.UI\umbraco\Views\Default.cshtml:line 115
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy()
   at System.Web.Mvc.WebViewPage.ExecutePageHierarchy()
   at System.Web.WebPages.WebPageBase.ExecutePageHierarchy(WebPageContext pageContext, TextWriter writer, WebPageRenderingBase startPage)
   at System.Web.Mvc.ViewResultBase.ExecuteResult(ControllerContext context)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult)
   at System.Web.Mvc.ControllerActionInvoker.InvokeActionResultFilterRecursive(IList`1 filters, Int32 filterIndex, ResultExecutingContext preContext, ControllerContext controllerContext, ActionResult actionResult)


```

**Comment by Shazwazza** [comment]
@nzdev  is this related to "This document is published but its url cannot be routed"  or a different issue that you are reporting? This seems unrelated.

**Comment by nzdev** [comment]
It may or not be related, happy to raise as a different issue.

**Comment by benbracedigital** [comment]
> Hi all - another question for everyone: For any items that become un-routable, are you using any of the built in 'special' umbraco route property types: `umbracoUrlAlias`, `umbracoUrlName`, `umbracoRedirectId` or `umbracoRedirect`?

None of those are used on our site that is affected.

**Comment by creativesuspects** [comment]
Not sure if it helps, but I noticed this issue also affects Multinode Treepicker data types where the root node is set based on an xpath query. I was using the following xpath query '$site/jobListing' for a 'Job Picker' property. When I tried to select a job using the picker I noticed the overlay was showing the full content tree instead of just the content underneath the 'jobListing' page (which is fallback behavior for a faulty xpath query). When I checked the 'jobListing' page it wasn't published due to the 'routing' issue. After rebuilding the cache the issue was gone and the picker worked as intended again.

Umbraco version: 8.11.1

**Comment by sbosell** [comment]
> Hi all - another question for everyone: For any items that become un-routable, are you using any of the built in 'special' umbraco route property types: `umbracoUrlAlias`, `umbracoUrlName`, `umbracoRedirectId` or `umbracoRedirect`?

The node that is being derouted has the special properties (I add them to all content nodes) but none of them are in use.

**Comment by Nicholas-Westby** [comment]
> Hi all - another question for everyone: For any items that become un-routable, are you using any of the built in 'special' umbraco route property types: `umbracoUrlAlias`, `umbracoUrlName`, `umbracoRedirectId` or `umbracoRedirect`?

Nope, we aren't using those properties.

**Comment by MartinThomasCW** [comment]
Nope, we aren't either.   We have umbracoUrlAlias defined, but not a single node uses it

**Comment by Shazwazza** [comment]
@orsinic wondering if you would be able to invite me to one of your Cloud projects exhibiting this issue? (shannon at umbraco dot com)

**Comment by Zweben** [comment]
> 
> 
> @Zweben
> 
> > I could probably invite you to the project if you think it would help. I'll also try to remember to turn on verbose logging next time I publish something to Live.
> 
> Yes it would certainly help me if you could invite me that would be great and also let me know if/when you enable that logging. It only needs to be enabled before and up to when you experience this issue and then can be disabled again.

I sent you an invite.

**Comment by Shazwazza** [comment]
@Zweben many thanks. Based on your above comments I noticed you mentioned this occurs mostly with the home page? I assume you mean the very first node in your tree? And when you've mentioned it occurs after a "deployment" - can you tell me what type of deployment specifically? A content deployment or a git/schema deployment? And any item(s) in particular? If I can know these answers I can try to replicate.

**Comment by kimschurmann** [comment]
> @Zweben many thanks. Based on your above comments I noticed you mentioned this occurs mostly with the home page? I assume you mean the very first node in your tree? And when you've mentioned it occurs after a "deployment" - can you tell me what type of deployment specifically? A content deployment or a git/schema deployment? And any item(s) in particular? If I can know these answers I can try to replicate.

Using Azure WebApps with deployment slots and using uSync to sync changes to production. First thought was that the swapping of disks when using deployment slots could invalidate the cache layer - but can't reproduce it. Another thought was that it could be uSync - but can't reproduce that either...

**Comment by Zweben** [comment]
> @Zweben many thanks. Based on your above comments I noticed you mentioned this occurs mostly with the home page? I assume you mean the very first node in your tree? And when you've mentioned it occurs after a "deployment" - can you tell me what type of deployment specifically? A content deployment or a git/schema deployment? And any item(s) in particular? If I can know these answers I can try to replicate.

In the past, this or a similar issue has happened on other nodes, but currently it seems to only affect the homepage. It is one of two root nodes currently, but it's node that has 99% of the content under it.

The issue seems to occur after doing a git/schema deployment from Development to Live in the Umbraco.io portal. After the deployment, the homepage will show a 404 error, and I will have to reload the database cache before it will display properly again.

**Comment by Zweben** [comment]
> 
> 
> @MartinThomasCW is your site an Umbraco Cloud site? If so, would it be possible to invite me as an admin to help investigate? (shannon at umbraco dot com). And/or if at all possible to enable this verbose logging in the serilog.config file: `<add key="serilog:minimum-level:override:Umbraco.Web.Routing.PublishedRouter" value="Debug" />` just so that it's enabled when this issue happens and then it can be turned off again. It's quite verbose so will log a lot. If that's possible could you email me those logs?

@Shazwazza I just did a deployment from Dev to Live that also contained the serilog config change to enable debug logging. It doesn't look like any errors occurred this time, though. Is it safe to leave debug logging on until our next deployment, to give it another shot?

**Comment by creativesuspects** [comment]
@Shazwazza I’ve been dealing with quite a few derouted node issues on a local dev environment. I will also enable debug logging and I’ll send you the logs when it happens again.

**Comment by Shazwazza** [comment]
@Zweben in your case, I can replicate issues when transferring schema. I don't get the home page as "document is published but its url cannot be routed" but instead on the child page "Who we..." when clicked I get a YSOD (I'm running with debug=true) in the web.config. There's a null reference exception coming from Your code in your content finder (i've obfuscated a few things):

```cs
NullReferenceException
at A.B.C.ChildSiteHelper.GetChildSiteSegmentFromUrl(Uri uri)
   at A.B.C.ChildSiteHelper.IsChildSiteRequest(Uri uri)
   at A.B.C.ChildSiteContentFinder.TryFindContent(PublishedRequest request)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Umbraco.Web.Routing.PublishedRouter.FindPublishedContent(PublishedRequest request)
   at Umbraco.Web.Routing.PublishedRouter.TryRouteRequest(PublishedRequest request)
   at Umbraco.Web.Routing.UrlProviderExtensions.DetectCollision(IContent content, String url, String culture, UmbracoContext umbracoContext, IPublishedRouter publishedRouter, ILocalizedTextService textService, UrlInfo& urlInfo)
   at Umbraco.Web.Routing.UrlProviderExtensions.<GetContentUrlsByCulture>d__1.MoveNext()
   at Umbraco.Web.Routing.UrlProviderExtensions.<GetContentUrls>d__0.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Umbraco.Web.Models.Mapping.ContentMapDefinition.Map(IContent source, ContentItemDisplay target, MapperContext context)
   at Umbraco.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, Type sourceType, MapperContext context)
   at Umbraco.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, MapperContext context)
   at Umbraco.Web.Editors.ContentController.MapToDisplay(IContent content)
   at lambda_method(Closure , Object , Object[] )
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_2.<GetExecutor>b__2(Object instance, Object[] methodParameters)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken 
```

The major problem is that you have **static** references to request based instances (as I can see in ChildSiteHelper... there's possibly others too), for example:

```cs
public static UmbracoContext _umbracoContext => Umbraco.Web.Composing.Current.UmbracoContext;
```

This absolutely cannot be done. The UmbracoContext is a request based lifetime and is based on the HttpContext. This is pinning this to the root of the app which is immediately stale and will hold all cache references to old stale caches (among many many other problems if you do this). This is described in the Common Pitfalls docs: https://our.umbraco.com/documentation/reference/common-pitfalls/#static-references-to-web-request-instances-such-as-umbracohelper

--- 

Can **everyone** make sure you are not pinning non-singletons to the root as statics. This will absolutely cause all sorts of problems like this one. 

My advice is to just stop using statics all together. It will make your code safer, more manageable, makes spaghetti code far more difficult to make, prevents memory leaks and prevents problems like there where a non-singleton object is pinned to the root as a singleton. Avoid statics and please use dependency injection it will make your code so much cleaner, easier to manage and you don't run into these problems.

If anyone exhibiting this issue can report back if this is their underlying problem it will be very helpful.

**Comment by sbosell** [comment]
@Shazwazza Related to your last message and mentioned in my opening text, we have a sync processing running every 30 minutes and are publishing the node that is having issues as well as its child  nodes.  I used the `RecurringTaskBase` example from the docs which looks something like:
`
```
 public class MenuSync : RecurringTaskBase
    {
        private IRuntimeState _runtime;
        private IProfilingLogger _logger;
        private IContentService _contentService;
       
        private IUmbracoContextFactory _umbracoContextFactory;
        private IScopeProvider _scopeProvider;
        public MenuSync(IBackgroundTaskRunner<RecurringTaskBase> runner, int delayBeforeWeStart, int howOftenWeRepeat, IRuntimeState runtime, IProfilingLogger logger, IContentService contentService, IUmbracoContextFactory umbracoContextFactory, IScopeProvider scopeProvider)
            : base(runner, delayBeforeWeStart, howOftenWeRepeat)
        {
            _runtime = runtime;
            _logger = logger;
            _contentService = contentService;
            _umbracoContextFactory = umbracoContextFactory;
            _scopeProvider = scopeProvider;
        }
```

The part that may be of question is that the only way I could get this to work with publishing and saving the content was to call `EnsureUmbracoContext` like below but perhaps that's not correct?  In my case ProcessLocation is calling Save or SaveAndPublish on each node that is being synced.  Is this correct?

  ```
public override bool PerformRun()
        {
            long count; 
            var locationsRoot = _contentService.GetById(1077);
            var locationsPublish = _contentService.GetPagedOfType(1073, 0, 1000, out count, null);

            using (var umb = _umbracoContextFactory.EnsureUmbracoContext())
            {
              
                    // do the publish stuff here 
                    foreach (var loc in locationsPublish)
                    {
                        ProcessLocation(loc);
                    }
            }
```

**Comment by Zweben** [comment]
> 
> 
> @Zweben in your case, I can replicate issues when transferring schema. I don't get the home page as "document is published but its url cannot be routed" but instead on the child page "Who we..." when clicked I get a YSOD (I'm running with debug=true) in the web.config. There's a null reference exception coming from Your code in your content finder (i've obfuscated a few things):
> 
> ```cs
> NullReferenceException
> at A.B.C.ChildSiteHelper.GetChildSiteSegmentFromUrl(Uri uri)
>    at A.B.C.ChildSiteHelper.IsChildSiteRequest(Uri uri)
>    at A.B.C.ChildSiteContentFinder.TryFindContent(PublishedRequest request)
>    at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
>    at Umbraco.Web.Routing.PublishedRouter.FindPublishedContent(PublishedRequest request)
>    at Umbraco.Web.Routing.PublishedRouter.TryRouteRequest(PublishedRequest request)
>    at Umbraco.Web.Routing.UrlProviderExtensions.DetectCollision(IContent content, String url, String culture, UmbracoContext umbracoContext, IPublishedRouter publishedRouter, ILocalizedTextService textService, UrlInfo& urlInfo)
>    at Umbraco.Web.Routing.UrlProviderExtensions.<GetContentUrlsByCulture>d__1.MoveNext()
>    at Umbraco.Web.Routing.UrlProviderExtensions.<GetContentUrls>d__0.MoveNext()
>    at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
>    at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
>    at Umbraco.Web.Models.Mapping.ContentMapDefinition.Map(IContent source, ContentItemDisplay target, MapperContext context)
>    at Umbraco.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, Type sourceType, MapperContext context)
>    at Umbraco.Core.Mapping.UmbracoMapper.Map[TTarget](Object source, MapperContext context)
>    at Umbraco.Web.Editors.ContentController.MapToDisplay(IContent content)
>    at lambda_method(Closure , Object , Object[] )
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_2.<GetExecutor>b__2(Object instance, Object[] methodParameters)
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken 
> ```
> 
> The major problem is that you have **static** references to request based instances (as I can see in ChildSiteHelper... there's possibly others too), for example:
> 
> ```cs
> public static UmbracoContext _umbracoContext => Umbraco.Web.Composing.Current.UmbracoContext;
> ```
> 
> This absolutely cannot be done. The UmbracoContext is a request based lifetime and is based on the HttpContext. This is pinning this to the root of the app which is immediately stale and will hold all cache references to old stale caches (among many many other problems if you do this). This is described in the Common Pitfalls docs: https://our.umbraco.com/documentation/reference/common-pitfalls/#static-references-to-web-request-instances-such-as-umbracohelper
> 
> Can **everyone** make sure you are not pinning non-singletons to the root as statics. This will absolutely cause all sorts of problems like this one.
> 
> My advice is to just stop using statics all together. It will make your code safer, more manageable, makes spaghetti code far more difficult to make, prevents memory leaks and prevents problems like there where a non-singleton object is pinned to the root as a singleton. Avoid statics and please use dependency injection it will make your code so much cleaner, easier to manage and you don't run into these problems.
> 
> If anyone exhibiting this issue can report back if this is their underlying problem it will be very helpful.

Oof, thank you @Shazwazza for identifying that. Let's just say I have learned quite a bit about .NET/C# since writing some of this code...

I believe I fell into that trap because I had extension methods in the class that tied into the other methods, but it's obvious that those need to be split out, and the dependencies injected instead. I'm going to do a search through our code-base for static references to any Umbraco objects, and try to get rid of all of them. You're a huge help!

**Comment by kimschurmann** [comment]
Any progress made on this? We are still stuck on 8.5.5 - any one knows if the problem will get worse if we upgrade?

**Comment by bergmania** [comment]
@kimschurmann as you can see from @Shazwazza's answers in this thread. We have still not found anything in Umbraco that leads to this directly. In the cases @Shazwazza have looked into so far, it has been custom code or packages with custom code that was the root cause. 

Did you check your solution do not suffer for the same issues as described in @Shazwazza's answers?

Regarding whether it will be worse if you upgrade, we currently don't think so.

**Comment by kimschurmann** [comment]
> @kimschurmann as you can see from @Shazwazza's answers in this thread. We have still not found anything in Umbraco that leads to this directly. In the cases @Shazwazza have looked into so far, it has been custom code or packages with custom code that was the root cause.
> 
> Did you check your solution do not suffer for the same issues as described in @Shazwazza's answers?
> 
> Regarding whether it will be worse if you upgrade, we currently don't think so.

We have a couple of places where we use Current.UmbracoContext - is that a problem too?

**Comment by bergmania** [comment]
>We have a couple of places where we use Current.UmbracoContext - is that a problem too?
>
>But it is not pinned as a static though!

That should be okay. You do not have any custom content finders? If you have, please ensure these are not introducing the issue.

**Comment by kimschurmann** [comment]
> > We have a couple of places where we use Current.UmbracoContext - is that a problem too?
> > But it is not pinned as a static though!
> 
> That should be okay. You do not have any custom content finders? If you have, please ensure these are not introducing the issue.

We only have one and it is very simple - and do not introduce this issue.

**Comment by sbosell** [comment]
Who in this thread has no custom packages. I don't and posted code and got zero response if that could be the issue.

**Comment by vaags** [comment]
For us this issue was really about missing languages under "culture and hostnames" on the root-node. The published document had a language that was not listed, and was therefore not being routed.

**Comment by Shazwazza** [comment]
Thanks @vaags, good find. 

If everyone can please check that this might be your issue and let us know that would be appreciated.

---
@kimschurmann `Current.UmbracoContext` should be ok so long as you are using it in a request thread (still best to use DI  wherever possible) but please definitely make sure you are never assigning the result of this to a static variable anywhere. This will cause problems.

---
I'll re-state my above comment in case it was overlooked:

> Can everyone make sure you are not pinning non-singletons to the root as statics. This will absolutely cause all sorts of problems like this one.

Example: **do not ever store the UmbracoContext or HttpContext into a static variable**

> My advice is to just stop using statics all together. It will make your code safer, more manageable, makes spaghetti code far more difficult to make, prevents memory leaks and prevents problems like there where a non-singleton object is pinned to the root as a singleton. Avoid statics and please use dependency injection it will make your code so much cleaner, easier to manage and you don't run into these problems.

> If anyone exhibiting this issue can report back if this is their underlying problem it will be very helpful.

---
@sbosell 

RE: https://github.com/umbraco/Umbraco-CMS/issues/9514#issuecomment-785946163 
Your usage of `using (var umb = _umbracoContextFactory.EnsureUmbracoContext())` is correct, but what are you doing with the `umb` instance? Does your `ProcessLocation` relying on Current.UmbracoContext?? You should be passing the actual UmbracoContext around and not relying on Current.UmbracoContext.  Since this is a background thread you should also be wrapping all of this in a custom Scope. So in `PerformRun` you should have this wrapped around everything:

```cs
using (var scope = _scopeProvider.CreateScope()) 
{
   // TODO: do stuff

   // Complete the scope, this completes the transaction
   // If an exception occurs anywhere above then this isn't called and the whole trans is rolled back.
   scope.Complete(); 
}
```

I was just re-reading all msgs here and revisiting this one: 
 https://github.com/umbraco/Umbraco-CMS/issues/9514#issuecomment-768255937 Are pages that de-route only the ones related to /locations? That would be interesting for 2 reasons: 

* Locations is related to the code above
* That would indicate it has some sort of relevance to the UmbracoVirtualNodeRouteHandler

**Comment by Zweben** [comment]
@Shazwazza:

> I'll re-state my above comment in case it was overlooked:
> 
> > Can everyone make sure you are not pinning non-singletons to the root as statics. This will absolutely cause all sorts of problems like this one.
> 
> Example: **do not ever store the UmbracoContext or HttpContext into a static variable**
> 
> > My advice is to just stop using statics all together. It will make your code safer, more manageable, makes spaghetti code far more difficult to make, prevents memory leaks and prevents problems like there where a non-singleton object is pinned to the root as a singleton. Avoid statics and please use dependency injection it will make your code so much cleaner, easier to manage and you don't run into these problems.
> 
> > If anyone exhibiting this issue can report back if this is their underlying problem it will be very helpful.

Reporting back with an update on my comment from 15 days ago. My code had static references to UmbracoContext because I didn't know better at the time. Although admittedly I haven't done a lot of work on the site in the past couple weeks, since removing those and switching everything to use dependency injection, I have not observed any further issues.

Just wondering, is it technically possible for the CMS to detect things like stale references to UmbracoContext, and provide users with a warning? Having a "common pitfalls" page is good, but I'd imagine that an active warning for things like that, perhaps in a health check, could do wonders for perceived stability.

**Comment by kimschurmann** [comment]
FYI: 

We took the risk and upgraded to 8.11.1 and made extensive test and monitoring on our test environment. We did not find any issues at all and have now upgraded in our production environment too. 

Although upgrading from 8.5.5 to 8.11.1 made content unavailable until we wiped the internal caches from the backoffice - this should probably be done automatically as a part of the upgrade?

**Comment by vaags** [comment]
> Although upgrading from 8.5.5 to 8.11.1 made content unavailable until we wiped the internal caches from the backoffice - this should probably be done automatically as a part of the upgrade?

I also exprerienced this. See #9378

**Comment by Shazwazza** [comment]
@Zweben 

Thanks for responding.

> Just wondering, is it technically possible for the CMS to detect things like stale references to UmbracoContext, and provide users with a warning? Having a "common pitfalls" page is good, but I'd imagine that an active warning for things like that, perhaps in a health check, could do wonders for perceived stability.

It's not. It could be done with static code analysis tools but not at runtime by the CMS.

---

I'm closing this issue now as there's been no further responses for my questions and some folks here have had this issue resolved by one or more of the solutions proposed in the above discussion.

**Comment by FransdeJong** [comment]
I ran into this issue on a site that went live yesterday.
There is one specific page as far as I know that is dropping from cache with the notification that the Url cannot be routed

@Shazwazza I checked the static context and we have nothing like that in place.
This page is a simple page that uses no route hijacking, no contentfinders or other magic solutions that would cause this to happen.

This is a cloud project so I'm able to add you or anyone else to the project if you want to investigate this further.
If you need any information please let me know.

**Comment by dawoe** [comment]
@Shazwazza @FransdeJong 

I noticed this as well on Umbraco Cloud several times after a deployment. I guess this happens when a doctype changes. Will keep monitoring it.

Dave

**Comment by FransdeJong** [comment]
@dawoe 
In this case I didn't push anything. It just drops from the cache.
Reloading memory cache fixes it.

**Comment by nathanwoulfe** [comment]
Apologies for resurrecting an old thread, but I'm noticing something which may be related - 

> This error just occurred right now.
> 
> ![image](https://user-images.githubusercontent.com/1896962/101559516-483d1000-398f-11eb-8cad-20492b2abe89.png)

Note the node name => Locations

Any time I try to create a node named Locations, I get a routing error:

`This document is published but its URL would collide with content /site.url/parentNodeName (id=1234)`

Doesn't matter where in the site I create the node, the error is always identical, referencing the parent node.

Might just be a coincidence, but it's an odd one if so...

@sbosell @Jay-umbr @FransdeJong @dawoe

**Comment by Nicholas-Westby** [comment]
@nathanwoulfe I wonder if it's related to the fact that your node name matches your template name (and your doctype name for that matter).

**Comment by nathanwoulfe** [comment]
@Nicholas-Westby nah, not in this case:
- Node name: Locations
- Node type: Super textpage (yup, I'm proud of naming that one)
- Template: Super textpage

Parent node is the same type and template, name is About.

Haven't tried creating Locations under other node types and using different node types, will do and see what happens...

**Comment by Nicholas-Westby** [comment]
@nathanwoulfe I see, so you ran into the same problem when you created a page called "Locations" that was using the doctype and template "Super textpage"? I could see it still being a problem unless you deleted the "Locations" template entirely (because the node name would still match **SOME** template, even though it isn't the template used on that node).

Though, your counter example of having a page called "About" with a doctype called "About" and a template called "About" seems to indicate the naming of the doctype and template are not an issue.

Also interesting that it happens anywhere in the tree. I can't remember if templates can be served up via subnodes, though I remember that at least they can be served up via the root (e.g., site.com/locations when Locations.cshtml exists).

**Comment by nathanwoulfe** [comment]
Ah, you're right - it will be catching Locations as an alternate template for the parent page, which is a bit odd since the Locations template isn't allowed for anything other than the Locations content type (will double check that, but that should be the case). 

Interestingly not an issue in v7, only surfaced as part of a V8 upgrade.

**Comment by jomehmet** [comment]
Got this error, almost all pages ended up in 404. 

Looking in the logs after this happened to us, I don't find any clear errors, but when I investigate the source code on github I wonder if there is a logical error. If you have a look in the snippet from the logs, it says `Resume document cache (reload:False)`
which is strange, since in the [source code](https://github.com/umbraco/Umbraco-CMS/blob/c0c9c50e2110a88afab800abbf0e4a6b6a08c62d/src/Umbraco.Infrastructure/Suspendable.cs#L45) we can se reload:True would trigger 
`pageRefresher?.RefreshAll()` that I excepect would refresh the cache, like when we go to Settings-> Published Status -> Reload Memory Cache

Wouldn't it help to let the method [SuspendDocumentCache ](https://github.com/umbraco/Umbraco-CMS/blob/c0c9c50e2110a88afab800abbf0e4a6b6a08c62d/src/Umbraco.Infrastructure/Suspendable.cs#L33)which is shown in the logs to be executed before ResumeDocumentCache to set s_tried=true

I can't figure out where this code is consumed from Github, but would apriciate if someone could verify if this is a logical error.

![image](https://user-images.githubusercontent.com/840606/165373625-17cea2a9-f1f0-4da5-b1c4-4438125db964.png)
